### PR TITLE
Fix sync encryption for Chromium 89

### PR DESCRIPTION
**Pull request by AlexeyBarabash**
This PR modifies the way how the passphrase is set for encryption/decryption according to Chromium 89 changes.

Resolves https://github.com/brave/brave-browser/issues/13657

## Submitter Checklist:

- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed).
- [ ] Requested a security/privacy review as needed.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
1. Connect to existing sync chain with code phrase
2. Expected to be connected

### CODE REVIEW

**Comment by AlexeyBarabash** [comment]
Need to rebase to the latest branch

**Comment by AlexeyBarabash** [comment]
Merging this because it blocks sync chain creation required for creating PR for https://github.com/brave/brave-browser/issues/13699 . Any found issue will be fixed in new PRs.

File: browser/resources/settings/brave_sync_page/brave_sync_page.js [hunk 1]
```diff
@@ -85,30 +84,24 @@ Polymer({
...
-  handleSyncPrefsChanged_: async function(syncPrefs) {
-    // Enforce encryption
-    if (this.syncStatus_ && !this.syncStatus_.firstSetupInProgress) {
-      if (!syncPrefs.encryptAllData) {
-        const syncCode = await this.braveBrowserProxy_.getSyncCode()
-        syncPrefs.encryptAllData = true;
-        syncPrefs.setNewPassphrase = true;
-        syncPrefs.passphrase = syncCode;
-        await this.browserProxy_.setSyncEncryption(syncPrefs)
-      } else if (syncPrefs.passphraseRequired) {
```
Comment (darkdh) [inline browser/resources/settings/brave_sync_page/brave_sync_page.js:106]: We still need to check `passphraseRequired` to decide whether to call `setDecryptionPassphrase` or `setEncryptionPassphrase` but not both.

File: browser/resources/settings/brave_sync_page/brave_sync_page.js [hunk 2]
```diff
@@ -39,6 +39,7 @@ Polymer({
      * @type {?SyncStatus}
      */
     syncStatus_: Object,
+    is_encryption_set_: Object,
```
Comment (darkdh) [inline browser/resources/settings/brave_sync_page/brave_sync_page.js:42]: This should be 
```   
{
  type: Boolean,
  value: false,
},
```
And naming convention should be `isEncryptionSet_`

File: browser/resources/settings/brave_sync_page/brave_sync_page.js [hunk 3]
```diff
@@ -85,30 +84,24 @@ Polymer({
...
-        const syncCode = await this.braveBrowserProxy_.getSyncCode()
-        syncPrefs.setNewPassphrase = false;
-        syncPrefs.passphrase = syncCode;
-        await this.browserProxy_.setSyncEncryption(syncPrefs)
-      }
+    // Cleanup is_encryption_set_ for re-enabling sync
+    if (!this.syncStatus_.isEngineInitialized &&
+      this.syncStatus_.firstSetupInProgress && this.is_encryption_set_) {
+      this.is_encryption_set_ = false
+    }
```
Comment (darkdh) [inline browser/resources/settings/brave_sync_page/brave_sync_page.js:94]: I don't think we need this if we provide default value correctly as I suggested above

File: browser/resources/settings/brave_sync_page/brave_sync_page.js [hunk 4]
```diff
@@ -85,30 +84,24 @@ Polymer({
...
+
+    // Both setEncryptionPassphrase and setDecryptionPassphrase need to have
+    // SyncService::IsEngineInitialized() true, see sync_service_crypto.cc
+    // We cannot rely on `firstSetupInProgress` because there is a gap when both
+    // `firstSetupInProgress` and `isEngineInitialized` are false
+    if (this.syncStatus_.isEngineInitialized && !this.is_encryption_set_) {
+      const syncCode = await this.braveBrowserProxy_.getSyncCode()
+      this.browserProxy_.setEncryptionPassphrase(syncCode);
```
Comment (darkdh) [inline browser/resources/settings/brave_sync_page/brave_sync_page.js:102]: `isEncryptionSet_` should be set from the return value of `setEncryptionPassphrase`, it is not always success.

File: browser/resources/settings/brave_sync_page/brave_sync_page.js [hunk 5]
```diff
@@ -85,30 +84,24 @@ Polymer({
...
+      this.browserProxy_.setDecryptionPassphrase(syncCode);
```
Comment (darkdh) [inline browser/resources/settings/brave_sync_page/brave_sync_page.js:103]: This is a promise, please add `await`

File: chromium_src/chrome/browser/ui/webui/settings/people_handler.cc [hunk 6]
```diff
@@ -9,7 +9,10 @@
   sync_status->SetBoolean(                \
       "firstSetupInProgress",             \
       service && !disallowed_by_policy && \
-          !service->GetUserSettings()->IsFirstSetupComplete());
+          !service->GetUserSettings()->IsFirstSetupComplete()); \
+  sync_status->SetBoolean(                \
+      "isEngineInitialized",             \
```
Comment (darkdh) [inline chromium_src/chrome/browser/ui/webui/settings/people_handler.cc:14]: We absolutely do not need this, `PeopleHandler::PushSyncPrefs` will check `IsEngineInitialized` before firing the event, that is why we should change encryption/decryption handling in `handleSyncStatus_`. We still need `handleSyncPrefsChanged_`

**Comment by darkdh** [comment]
I pushed a commit(1a643b629c2637dd735401403efff11cd55e4c98) to cr89 directly

**Pull request merged**